sky/serve/load_balancing_policies.py

```diff
@@ -11,48 +11,57 @@
 logger = sky_logging.init_logger(__name__)
 
 
+def _request_repr(request: 'fastapi.Request') -> str:
+    return ('<Request '
+            f'method="{request.method}" '
+            f'url="{request.url}" '
+            f'headers={dict(request.headers)} '
+            f'query_params={dict(request.query_params)}>')
+
+
 class LoadBalancingPolicy:
     """Abstract class for load balancing policies."""
 
-    def __init__(self) -> None:
-        self.ready_replicas: List[str] = []
-
     def set_ready_replicas(self, ready_replicas: List[str]) -> None:
         raise NotImplementedError
 
+    def select_replica(self, request: 'fastapi.Request') -> Optional[str]:
+        replica = self._select_replica(request)
+        if replica is not None:
+            logger.info(f'Selected replica {replica} '
+                        f'for request {_request_repr(request)}')
+        else:
+            logger.warning('No replica selected for request '
+                           f'{_request_repr(request)}')
+        return replica
+
     # TODO(tian): We should have an abstract class for Request to
     # compatible with all frameworks.
-    def select_replica(self, request: 'fastapi.Request') -> Optional[str]:
+    def _select_replica(self, request: 'fastapi.Request') -> Optional[str]:
         raise NotImplementedError
 
 
 class RoundRobinPolicy(LoadBalancingPolicy):
     """Round-robin load balancing policy."""
 
-    def __init__(self, *args, **kwargs) -> None:
-        super().__init__(*args, **kwargs)
+    def __init__(self) -> None:
+        self.ready_replicas: List[str] = []
```

Comment:
why do we move the ready_replicas here? I assume most of the underlying load balancer policy should need the `ready_replicas` information.

Reply:
It is to avoid redundant information, i.e. if the customized load balancer have a map from ready replicas to their properties, this list is redundant.

Reply:
Oh nvm. I just realized the load balancer also need this information for maintaining client to corresponding ready replica urls. Changed back :)